src/Interfaces/FwupdManager.vala

```diff
@@ -0,0 +1,317 @@
+                case "Checksum":
+                    release.checksum = val.get_string ();
+                    break;
+                case "Vendor":
+                    release.vendor = val.get_string ();
+                    break;
+                case "Size":
+                    release.size = val.get_uint64 ();
+                    break;
+                case "License":
+                    release.license = val.get_string ();
+                    break;
+                case "TrustFlags":
+                    release.flag = (ReleaseFlag) val.get_uint64 ();
+                    break;
+                case "InstallDuration":
+                    release.install_duration = val.get_uint32 ();
+                    break;
+                case "Uri":
+                    release.uri = val.get_string ();
+                    break;
+                default:
+                    break;
+            }
+        });
+
+        return release;
+    }
+
+    private string get_path (string uri) {
+        var parts = uri.split ("/");
+        string file_path = parts[parts.length - 1];
+        return Path.build_path (Path.DIR_SEPARATOR_S, Environment.get_tmp_dir (), file_path);
+    }
+
+    public async string? download_file (Device device, string uri) {
+        var path = get_path (uri);
+
+        File server_file = File.new_for_uri (uri);
+        File local_file = File.new_for_path (path);
+
+        if (FileUtils.test (path, FileTest.IS_REGULAR)) {
+            return path;
+        }
+
+        bool result;
+        try {
+            result = yield server_file.copy_async (local_file, FileCopyFlags.OVERWRITE, Priority.DEFAULT, null, (current_num_bytes, total_num_bytes) => {});
+        } catch (Error e) {
+            on_device_error (device, "Could not download file: %s".printf (e.message));
+            return null;
+        }
+
+        if (!result) {
+            on_device_error (device, "Download of %s was not succesfull".printf (uri));
+            return null;
+        }
+
+        return path;
+    }
+
+    public async bool install (Device device, string path) {
+        try {
+            // https://github.com/fwupd/fwupd/blob/c0d4c09a02a40167e9de57f82c0033bb92e24167/libfwupd/fwupd-client.c#L2045
+            var options = new GLib.HashTable<string, Variant> (str_hash, str_equal);
+            options.insert ("reason", new Variant.string ("user-action"));
+            options.insert ("filename", new Variant.string (path));
+            options.insert ("allow-older", new Variant.boolean (true));
+            options.insert ("allow-reinstall", new Variant.boolean (true));
+            options.insert ("no-history", new Variant.boolean (true));
+
+            var fd = ro_fd (path);
+            var handle = new UnixInputStream (fd, true);
+
+            yield fwupd.install (device.id, handle, options);
+        } catch (Error e) {
+            warning ("Could not connect to fwupd interface: %s", e.message);
```

Comment:
This warning message is too generic, we're already connected to the fwupd interface, this is a failure to install a firmware update, right?